Mathlib/MeasureTheory/MeasurableSpace/CountablyGenerated.lean

```diff
@@ -99,6 +117,140 @@ instance [MeasurableSpace α] [CountablyGenerated α] [MeasurableSpace β] [Coun
+def map_nat_bool_of_countablyGenerated [MeasurableSpace α] [CountablyGenerated α] (x : α) (n : ℕ) :
+    Bool := x ∈ natGeneratingSequence α n
+
+theorem measurable_map_nat_bool_of_countablyGenerated [MeasurableSpace α] [CountablyGenerated α] :
+    Measurable (map_nat_bool_of_countablyGenerated α) := by
+  rw [measurable_pi_iff]
+  refine fun n ↦ measurable_to_bool ?_
+  simp only [preimage, mem_singleton_iff, map_nat_bool_of_countablyGenerated,
+    Bool.decide_iff, setOf_mem_eq]
+  apply measurableSet_natGeneratingSequence
+
+theorem injective_map_nat_bool_of_countablyGenerated [MeasurableSpace α] [CountablyGenerated α]
+    [SeparatesPoints α] : Injective (map_nat_bool_of_countablyGenerated α) := by
+  intro x y hxy
+  rw [← generateFrom_natGeneratingSequence α] at *
+  apply separating_of_generateFrom (range (natGeneratingSequence _))
+  rintro - ⟨n, rfl⟩
+  rw [← decide_eq_decide]
+  exact congr_fun hxy n
+
+/-- If a measurable space is countably generated and separates points, it is measure equivalent
+to some some subset of the Cantor space `ℕ → Bool` (equipped with the product sigma algebra).
+Note: `s` need not be measurable, so this map need not be a `MeasurableEmbedding` to
+the Cantor Space. -/
+theorem measurableEquiv_nat_bool_of_countablyGenerated [MeasurableSpace α]
+    [CountablyGenerated α] [SeparatesPoints α] :
+    ∃ s : Set (ℕ → Bool), Nonempty (α ≃ᵐ s) := by
+  use range (map_nat_bool_of_countablyGenerated α), Equiv.ofInjective _ $
+    injective_map_nat_bool_of_countablyGenerated _,
+    Measurable.subtype_mk $ measurable_map_nat_bool_of_countablyGenerated _
+  simp_rw [← generateFrom_natGeneratingSequence α]
+  apply measurable_generateFrom
+  rintro _ ⟨n, rfl⟩
+  rw [← Equiv.image_eq_preimage _ _]
+  refine' ⟨{y | y n}, by measurability, _⟩
+  rw [← Equiv.preimage_eq_iff_eq_image]
+  simp[map_nat_bool_of_countablyGenerated]
+
+/-- If a measurable space admits a countable sequence of measurable sets separating points,
+it admits a measurable injection into the Cantor space `ℕ → Bool`
+(equipped with the product sigma algebra). -/
+theorem measurable_injection_nat_bool_of_hasCountableSeparatingOn [MeasurableSpace α]
+    [HasCountableSeparatingOn α MeasurableSet univ] :
+    ∃ f : α → ℕ → Bool, Measurable f ∧ Injective f := by
+  rcases exists_countablyGenerated_le_of_hasCountableSeparatingOn α with ⟨m', _, _, m'le⟩
+  refine' ⟨map_nat_bool_of_countablyGenerated α, _, injective_map_nat_bool_of_countablyGenerated _⟩
+  exact (measurable_map_nat_bool_of_countablyGenerated _).mono m'le (by rfl)
+
+variable {α}
+
+--TODO: Make this an instance
+theorem measurableSingletonClass_of_hasCountableSeparatingOn
+    [MeasurableSpace α] [HasCountableSeparatingOn α MeasurableSet univ] :
+    MeasurableSingletonClass α := by
+  rcases measurable_injection_nat_bool_of_hasCountableSeparatingOn α with ⟨f, fmeas, finj⟩
+  refine' ⟨fun x ↦ _⟩
+  rw [← finj.preimage_image {x}, image_singleton]
+  exact fmeas $ MeasurableSet.singleton _
```

Comment:
This theorem is also about the function for which you added a def. Here is a proposition for a proof that uses it directly:
```suggestion
    MeasurableSingletonClass α where
  measurableSet_singleton x := by
    rcases exists_countablyGenerated_le_of_hasCountableSeparatingOn α with ⟨m', _, _, m'le⟩
    rw [← (injective_map_nat_bool_of_countablyGenerated _).preimage_image {x}, image_singleton]
    exact (measurable_map_nat_bool_of_countablyGenerated α).mono m'le le_rfl
      $ MeasurableSet.singleton _
```

Reply:
The original proof seems more straightforward to me, mostly because the proposed one repeats a lot of the proof of `measurable_injection_nat_bool_of_hasCountableSeparatingOn` anyways.

Reply:
The rest of the suggestions have been implemented, thanks again!

Reply:
Fine by me. I agree that my proof repeats the other one quite a bit.